dev/DynamicDependency/PackageDependencyManager.cpp

```diff
@@ -0,0 +1,171 @@
+
+    auto lock{ std::unique_lock<std::mutex>(g_lock) };
+
+    g_packageDependencies.push_back(packageDependency);
+
+    auto id{ wil::make_process_heap_string(packageDependency.Id().c_str()) };
+    *packageDependencyId = id.release();
+}
+
+void MddCore::PackageDependencyManager::CreatePackageDependency(
+    const MddCore::PackageDependency& packageDependency,
+    _Outptr_result_maybenull_ PWSTR* packageDependencyId)
+{
+    // Make a new PackageDependency object with the definition parts of the provided one
+    CreatePackageDependency(packageDependency.User(), packageDependency.PackageFamilyName().c_str(),
+        packageDependency.MinVersion(), packageDependency.Architectures(),
+        packageDependency.LifetimeKind(), packageDependency.LifetimeArtifact().c_str(), packageDependency.Options(),
+        packageDependencyId);
+}
+
+void MddCore::PackageDependencyManager::DeletePackageDependency(
+    _In_ PCWSTR packageDependencyId)
+{
+    if (!packageDependencyId)
+    {
+        return;
+    }
+
+    auto lock{ std::unique_lock<std::mutex>(g_lock) };
+
+    for (size_t index=0; index < g_packageDependencies.size(); ++index)
+    {
+        const auto& packageDependency{ g_packageDependencies[index] };
+        if (CompareStringOrdinal(packageDependency.Id().c_str(), -1, packageDependencyId, -1, TRUE) == CSTR_EQUAL)
+        {
+            g_packageDependencies.erase(g_packageDependencies.begin() + index);
+            break;
+        }
+    }
+
+    MddCore::DataStore::Delete(packageDependencyId);
+}
+
+const MddCore::PackageDependency* MddCore::PackageDependencyManager::GetPackageDependency(
+    _In_ PCWSTR packageDependencyId)
+{
+    // Find it (if we can)
+    auto packageDependency{ GetPackageDependencyInMemory(packageDependencyId) };
+    if (!packageDependency)
+    {
+        packageDependency = GetPackageDependencyInDataStore(packageDependencyId);
+    }
+    return packageDependency;
+}
+
+const MddCore::PackageDependency* MddCore::PackageDependencyManager::GetPackageDependencyInMemory(
+    _In_ PCWSTR packageDependencyId)
+{
+    // Check the in-memory list
+    for (size_t index=0; index < g_packageDependencies.size(); ++index)
+    {
+        const auto& packageDependency{ g_packageDependencies[index] };
+        if (CompareStringOrdinal(packageDependency.Id().c_str(), -1, packageDependencyId, -1, TRUE) == CSTR_EQUAL)
+        {
+            // Gotcha!
+            return &packageDependency;
+        }
+    }
+
+    // Not found
+    return nullptr;
+}
+
+/// @warning This method assumes \c packageDependencyId is not in the in-memory list
+const MddCore::PackageDependency* MddCore::PackageDependencyManager::GetPackageDependencyInDataStore(
+    _In_ PCWSTR packageDependencyId)
+{
+    auto packageDependency{ MddCore::DataStore::Load(packageDependencyId) };
+    if (!packageDependency)
+    {
+        return nullptr;
+    }
+
+    // We're guaranteed it's not in the in-memory list at this point
```

Comment:
this is an assumption, not a guarantee?  If it's in the list, how bad is that?

Reply:
Line 139 is redundant and can be removed.

FTI the warning @ line 129 dictates this is a prerequisite.

It that requirement was broken we could have entries in g_packageDependencies with duplicate unique ids - causing things like looking for a match to stop at the first and never sees 2+ dupe, etc. The caller only calls this packageDependencyId's not found in memory, but that prereq may not be clear to future developers hence line 129.